### PR TITLE
Reduce training overhead from GPU to CPU sync

### PR DESCRIPTION
**Pull request by workofart**
## Description
- Add explicit `eval()` for the option to force MLX's lazy parameter/state updates at optimizer-step boundaries. Previously, we were accumulating lazy ops across multiple steps, and this could potentially result in bursty / unexpected sync between GPU and CPU.
- Defer `clip_grad_norm` to work on GPU until absolutely necessary reporting/logging steps then move the grad norm back to CPU
- Decouple reporting from optimizer stepping. They can run separately. Previously, they were dependent.
- Move `max_grad_norm` from optimizer config to just general training config. This behavior is similar to PyTorch.

## Tests

| Change | Observed gain |
|---|---|
| Decouple reporting from optimizer stepping. | Better maintainability and determinism |
| Keep clipping behavior with max_grad_norm=1.0. The branch still clips gradients, but avoids scalar logging/sync cost when the value is not reported. | Correctness-preserving perf change |

For explicit `eval()` for materializing at predictable `Optimizer.step()`:

| Case | Explicit eval | Steps/s | Delta vs disabled | Note |
|---|---|---|---|---|
| report_every_steps=1 | enabled | 21.63 | -2.34% | Slightly slower when every step already reports/syncs. |
| report_every_steps=1 | disabled | 22.15 | baseline | No delayed sync buildup. |
| report_every_steps=100 | enabled | 21.31 | +12.1% | Clear win; lazy work is forced at optimizer-step boundaries. |
| report_every_steps=100 | disabled | 18.73 | baseline | Optimizer step looks cheap, but deferred work hits later. |

### CODE REVIEW

File: autograd/optim.py [hunk 1]
```diff
@@ -365,11 +373,11 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
     def step(self) -> None:
```
Comment (workofart) [inline autograd/optim.py:373]: Making this step() function truly abstract, just like PyTorch. And putting the specific operations into the optimizer subclasses for clarity.
